tsdb/engine/tsm1/encoding.go

```diff
@@ -111,26 +111,45 @@ func (a Values) Merge(b Values) Values {
 		return a
 	}
 
-	var i, j int
-	for ; i < len(a) && j < len(b); i++ {
-		av, bv := a[i].UnixNano(), b[j].UnixNano()
+	for i := 0; i < len(a) && len(b) > 0; i++ {
+		av, bv := a[i].UnixNano(), b[0].UnixNano()
+		// Value in a is greater than B, we need to merge
 		if av > bv {
-			a[i], b[j] = b[j], a[i]
+			// Save value in a
+			temp := a[i]
+
+			// Overwrite a with b
+			a[i] = b[0]
+
+			// Slide all values of b down 1
+			copy(b, b[1:])
+			b = b[:len(b)-1]
```

Comment:
It doesn't actually allocate in this loop.  It's just updating the slice length and shifting values.  The underlying storage does not change until L151 where it need to append the rest of b onto a.

I pushed up a benchmark to verify.

```
$ go test -run non -bench BenchmarkValues_Merge -benchmem -v ./tsdb/engine/tsm1/
PASS
BenchmarkValues_Merge-8	  100000	     16167 ns/op	   32769 B/op	       1 allocs/op
```

Removing the append on L151
```
$ go test -run non -bench BenchmarkValues_Merge -benchmem -v ./tsdb/engine/tsm1/
PASS
BenchmarkValues_Merge-8	  300000	      5178 ns/op	       0 B/op	       0 allocs/op
```